File: Trees/LC_105.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.30

// This solution is from Neetcode.io, I haven't written my own version using only index to indicate the range
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int index = 0;
        return build(preorder, inorder, index, 0, inorder.size() - 1);
    }
private:
    TreeNode* build(vector<int>& preorder, vector<int>& inorder, int& index, int i, int j) {
        if (i > j) {
            return NULL;
        }
        
        TreeNode* root = new TreeNode(preorder[index]);
        
        int split = 0;
        for (int i = 0; i < inorder.size(); i++) {
            if (preorder[index] == inorder[i]) {
                split = i;
                break;
            }
        }
        index++;
        
        root->left = build(preorder, inorder, index, i, split - 1);
        root->right = build(preorder, inorder, index, split + 1, j);
        
        return root;
    }
};

// reconstruct vector is not efficient, see the better solution 
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return build(preorder, inorder);
    }
private:
    TreeNode* build(vector<int>& preorder, vector<int>& inorder){
        if (preorder.size() == 0 || inorder.size() == 0){
            return nullptr;
        }
        int root_val = preorder[0];
        TreeNode* root = new TreeNode(root_val);
        int pos = 0;
        while (inorder[pos] != root_val){
            pos += 1;
        }
        vector<int> left_preorder, left_inorder;
        for (int i = 0; i < pos; i++){
            left_preorder.push_back(preorder[1 + i]);
            left_inorder.push_back(inorder[i]);
        }
        root->left = build(left_preorder, left_inorder);
        vector<int> right_preorder, right_inorder;
        for (int i = pos + 1; i < inorder.size(); i++){
            right_preorder.push_back(preorder[i]);
            right_inorder.push_back(inorder[i]);
        }
        root->right = build(right_preorder, right_inorder);
        return root;
    }
};

File: Trees/LC_105.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.30

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        # first node in preorder is always going to be root
        # find the location of root in the inorder, then the previous part
        # is left subtree, and the following part is the right subtree!
        
        def build(preorder, inorder):
            if not preorder or not inorder:
                return None
            root_val = preorder[0]
            root = TreeNode(root_val)
            idx = 0
            # find the location of root in the inorder array
            while inorder[idx] != root_val:
                idx += 1
            # build left subtree
            root.left = build(preorder[1:1+idx], inorder[:idx])
            # build right subtree
            root.right = build(preorder[1+idx:], inorder[idx+1:])
            
            return root
        return build(preorder, inorder)